centrifuge/src/main/java/io/github/centrifugal/centrifuge/Subscription.java

```diff
@@ -41,12 +46,38 @@ SubscriptionEventListener getListener() {
         this.listener = listener;
     }
 
+    public long getLastOffset() {
+        return lastOffset;
+    }
+
+    public void setLastOffset(long lastOffset) {
+        this.lastOffset = lastOffset;
+    }
+
+    public String getLastEpoch() {
+        return lastEpoch;
+    }
+
+    public void setLastEpoch(String lastEpoch) {
+        this.lastEpoch = lastEpoch;
+    }
+
     void moveToUnsubscribed() {
         this.state = SubscriptionState.UNSUBSCRIBED;
     }
 
     void moveToSubscribeSuccess(Protocol.SubscribeResult result) {
         this.state = SubscriptionState.SUBSCRIBED;
+        this.lastEpoch = result.getEpoch();
+        this.lastOffset = result.getOffset();
+        this.setRecover(result.getRecoverable());
+
+        for (Protocol.Publication publication: result.getPublicationsList()) {
+            PublishEvent publishEvent = new PublishEvent();
+            publishEvent.setData(publication.toByteArray());
+            this.listener.onPublish(this, publishEvent);
+        }
+
         SubscribeSuccessEvent event = new SubscribeSuccessEvent();
```

Comment:
`SubscribeSuccessEvent` should be extended with a coupled of fields: `Resubscribed` and `Recovered`. By analogy [with this for example](https://github.com/centrifugal/centrifuge-go/blob/master/subscription.go#L14). Let me explain why it's important. Without those fields developer can't understand whether connection recovered all publications after reconnect or not. Since Centrifuge/Centrifugo keep a sliding window of fixed size of publications in history client can reconnect too late: after messages in history expired or evicted by size. In case of successful recovery server returns a flag `recovered`: we need to pass it to `SubscribeSuccessEvent` to let user-level code to make required actions (load messages from database or sth). Since `SubscribeSuccessEvent` also called on first subscribe (where recovery not used) we also provide `Resubscribed` field to distinguish between cases where we used recovery and did not use it.

Reply:
@FZambia hi, I passed recovered and resubscrible to SubcribeSuccessEvent. You can check and comment for me. tks